archive/untar_stream.ts

```diff
@@ -0,0 +1,235 @@
+ * decompression which must be done before expanding the archive.
+ *
+ * @example
+ * ```ts
+ * import { UnTarStream } from "@std/archive/untar-stream";
+ *
+ * let fileWriter: WritableStreamDefaultWriter | undefined;
+ * for await (
+ *   const entry of (await Deno.open('./out.tar.gz'))
+ *     .readable
+ *     .pipeThrough(new DecompressionStream('gzip'))
+ *     .pipeThrough(new UnTarStream())
+ * ) {
+ *   if (entry.type === "header") {
+ *     fileWriter?.close();
+ *     fileWriter = (await Deno.create(entry.pathname)).writable.getWriter();
+ *   } else await fileWriter!.write(entry.data);
+ * }
+ * ```
+ */
+export class UnTarStream
+  implements TransformStream<Uint8Array, TarStreamChunk> {
+  #readable: ReadableStream<TarStreamChunk>;
+  #writable: WritableStream<Uint8Array>;
+  #gen: AsyncGenerator<Uint8Array>;
+  constructor() {
+    const { readable, writable } = new TransformStream<
+      Uint8Array,
+      Uint8Array
+    >();
+    this.#readable = ReadableStream.from(this.#untar());
+    this.#writable = writable;
+
+    this.#gen = async function* () {
+      const buffer: Uint8Array[] = [];
+      for await (
+        const chunk of readable.pipeThrough(new FixedChunkStream(512))
+      ) {
+        if (chunk.length !== 512) {
+          throw new Error("Tarball has an unexpected number of bytes");
+        }
+
+        buffer.push(chunk);
+        if (buffer.length > 2) yield buffer.shift()!;
+      }
+      if (buffer.length < 2) {
+        throw new Error("Tarball was too small to be valid");
+      }
+      if (!buffer.every((value) => value.every((x) => x === 0))) {
+        throw new Error("Tarball has invalid ending");
+      }
+    }();
+  }
+
+  async *#untar(): AsyncGenerator<TarStreamChunk> {
+    const decoder = new TextDecoder();
+    while (true) {
+      const { done, value } = await this.#gen.next();
+      if (done) break;
+
+      // Validate Checksum
+      const checksum = value.slice();
+      checksum.set(new Uint8Array(8).fill(32), 148);
+      if (
+        checksum.reduce((x, y) => x + y) !==
+          parseInt(decoder.decode(value.slice(148, 156 - 2)), 8)
+      ) throw new SyntaxError("Tarball header failed to pass checksum");
+
+      // Decode Header
+      let header: OldStyleFormat | PosixUstarFormat = {
+        name: decoder.decode(value.slice(0, 100)).replaceAll("\0", ""),
+        mode: decoder.decode(value.slice(100, 108 - 2)),
+        uid: decoder.decode(value.slice(108, 116 - 2)),
+        gid: decoder.decode(value.slice(116, 124 - 2)),
+        size: parseInt(decoder.decode(value.slice(124, 136)).trimEnd(), 8),
+        mtime: parseInt(decoder.decode(value.slice(136, 148 - 1)), 8),
+        checksum: decoder.decode(value.slice(148, 156 - 2)),
+        typeflag: decoder.decode(value.slice(156, 157)),
+        linkname: decoder.decode(value.slice(157, 257)).replaceAll(
+          "\0",
+          "",
+        ),
+        pad: value.slice(257),
```

Comment:
`checksum` and `pad` don't need to be exposed.

Reply:
Pad should arguably be exposed as it may contain additional information future versions may implement in that's backwards compatible.

Reply:
The new pax format is incredibly extensible, BSD tar has already adopted pax as the default output format, and the same change is planned for GNU tar, so I doubt ustar will receive any additions in the future. Even if there are new fields they can be added at any time, but removing `pad` or shrinking its size would be a breaking change.